File: python/src/sentinel_shield/protocol.py
"""SENTINEL Shield wire protocol (Python reference).

A newline-delimited JSON ("JSONL") protocol, identical to the TypeScript
reference implementation in ``src/shield/protocol.ts``. Every request carries a
correlation ``id`` echoed by the matching response, so a single connection can
multiplex concurrent requests.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional

PROTOCOL_VERSION = 1


@dataclass
class ToolCall:
    """A tool call as presented to the Shield."""

    tool: str
    args: Optional[Dict[str, Any]] = None
    text: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        out: Dict[str, Any] = {"tool": self.tool}
        if self.args is not None:
            out["args"] = self.args
        if self.text is not None:
            out["text"] = self.text
        return out


@dataclass
class Verdict:
    """The Shield's deterministic decision about a tool call."""

    verdict: str  # "allow" | "warn" | "block"
    risk: str
    score: int
    matches: List[Dict[str, Any]] = field(default_factory=list)
    allowed: bool = True

    @property
    def blocked(self) -> bool:
        return self.verdict == "block"

    @classmethod
    def from_response(cls, msg: Dict[str, Any]) -> "Verdict":
        return cls(
            verdict=msg["verdict"],
            risk=msg["risk"],
            score=msg["score"],
            matches=msg.get("matches", []),
            allowed=msg.get("allowed", msg["verdict"] != "block"),
        )


def encode(msg: Dict[str, Any]) -> bytes:
    """Encode a message as a single newline-terminated JSON line."""
    return (json.dumps(msg, separators=(",", ":")) + "\n").encode("utf-8")


class LineDecoder:
    """Stateful newline-delimited JSON decoder."""

    def __init__(self) -> None:
        self._buffer = b""

    def push(self, chunk: bytes) -> List[Dict[str, Any]]:
        self._buffer += chunk
        out: List[Dict[str, Any]] = []
        while b"\n" in self._buffer:
            line, self._buffer = self._buffer.split(b"\n", 1)
            line = line.strip()
            if line:
                out.append(json.loads(line.decode("utf-8")))
        return out


class ShieldError(RuntimeError):
    """Raised when the Shield returns an error response."""


class ShieldBlocked(RuntimeError):
    """Raised when a protected operation is blocked by the Shield."""

    def __init__(self, verdict: Verdict, tool: str) -> None:
        super().__init__(f"Shield blocked tool '{tool}' (risk={verdict.risk}, score={verdict.score})")
        self.verdict = verdict
        self.tool = tool

File: src/shield/protocol.ts
// ---------------------------------------------------------------------------
// SENTINEL Shield Protocol (v0.3.0, S1)
// ---------------------------------------------------------------------------
// A tiny, language-agnostic, newline-delimited JSON ("JSONL") protocol spoken
// between an agent and the out-of-band Shield process over a Unix domain
// socket or TCP. Any agent in any language can connect; the heavy safety logic
// lives in the Shield, not in the agent's process.
//
// Every request carries a correlation `id` that the matching response echoes,
// so a client can multiplex concurrent requests over a single connection.
// ---------------------------------------------------------------------------

import type { RiskLevel } from '../kernel/types.js';
import type { FirewallVerdict, PatternMatch } from '../firewall/firewall.js';
import type { KillMode, ForensicsSnapshot } from '../exec/killswitch.js';

export const SHIELD_PROTOCOL_VERSION = 1;

/** A tool call as presented to the Shield (mirrors firewall.ToolCall). */
export interface ShieldToolCall {
  tool: string;
  args?: Record<string, unknown>;
  text?: string;
}

// ---- Client → Shield requests ---------------------------------------------

export interface HelloRequest {
  type: 'hello';
  protocol: number;
  agent: string;
}

export interface ScanRequest {
  type: 'scan';
  id: string;
  call: ShieldToolCall;
}

export interface BeginRequest {
  type: 'begin';
  id: string;
  description: string;
}

export interface CompleteRequest {
  type: 'complete';
  id: string;
  opId: string;
}

export interface KillRequest {
  type: 'kill';
  id: string;
  /** Target agent; defaults to the calling session's agent. */
  agent?: string;
  mode?: KillMode;
  reason?: string;
}

export interface StatusRequest {
  type: 'status';
  id: string;
}

export interface PingRequest {
  type: 'ping';
  id: string;
}

export type ShieldRequest =
  | HelloRequest
  | ScanRequest
  | BeginRequest
  | CompleteRequest
  | KillRequest
  | StatusRequest
  | PingRequest;

// ---- Shield → Client responses --------------------------------------------

export interface WelcomeResponse {
  type: 'welcome';
  shieldVersion: string;
  protocol: number;
  policy: string;
  sessionId: string;
  /** Opaque per-session token; revoked on kill or watchdog fire. */
  token: string;
}

export interface VerdictResponse {
  type: 'verdict';
  id: string;
  verdict: FirewallVerdict;
  risk: RiskLevel;
  score: number;
  matches: PatternMatch[];
  /** Convenience: true when the call may proceed (verdict !== 'block'). */
  allowed: boolean;
}

export interface OkResponse {
  type: 'ok';
  id: string;
  opId?: string;
}

export interface KilledResponse {
  type: 'killed';
  id: string;
  snapshot: ForensicsSnapshot;
}

export interface StatusResponse {
  type: 'status';
  id: string;
  shieldVersion: string;
  policy: string;
  uptimeMs: number;
  killSwitch: 'armed' | 'fired';
  agents: Array<{ agent: string; status: string; operations: number }>;
  stats: { allowed: number; warned: number; blocked: number };
}

export interface PongResponse {
  type: 'pong';
  id: string;
}

export interface ErrorResponse {
  type: 'error';
  id?: string;
  message: string;
}

/** Pushed (unsolicited) when a session's token is revoked. */
export interface RevokedNotice {
  type: 'revoked';
  reason: string;
}

export type ShieldResponse =
  | WelcomeResponse
  | VerdictResponse
  | OkResponse
  | KilledResponse
  | StatusResponse
  | PongResponse
  | ErrorResponse
  | RevokedNotice;

// ---- Line framing ----------------------------------------------------------

/** Encode a message as a single newline-terminated JSON line. */
export function encodeMessage(msg: ShieldRequest | ShieldResponse): string {
  return JSON.stringify(msg) + '\n';
}

/**
 * Stateful newline-delimited JSON decoder. Feed it raw chunks; it yields fully
 * parsed messages and buffers any partial trailing line.
 */
export class LineDecoder<T> {
  private buffer = '';

  push(chunk: string): T[] {
    this.buffer += chunk;
    const out: T[] = [];
    let nl = this.buffer.indexOf('\n');
    while (nl >= 0) {
      const line = this.buffer.slice(0, nl).trim();
      this.buffer = this.buffer.slice(nl + 1);
      if (line.length > 0) {
        out.push(JSON.parse(line) as T);
      }
      nl = this.buffer.indexOf('\n');
    }
    return out;
  }
}
